src/devices/machine/spifi3.h

```diff
@@ -0,0 +1,465 @@
+		ARB_DESKEW_WAIT,
+
+		// Send/receive byte
+		SEND_WAIT_SETTLE,
+		SEND_WAIT_REQ_0,
+		RECV_WAIT_REQ_1,
+		RECV_WAIT_SETTLE,
+		RECV_WAIT_REQ_0
+	};
+
+	enum
+	{
+		STATE_MASK = 0x00ff,
+		SUB_SHIFT = 8,
+		SUB_MASK = 0xff00
+	};
+
+	enum
+	{
+		BUS_BUSY,
+		BUS_FREE_WAIT,
+		BUS_FREE
+	};
+
+	enum dma_direction
+	{
+		DMA_NONE,
+		DMA_IN,
+		DMA_OUT
+	};
+	static inline bool dma_command(dma_direction current_direction)
+	{
+		return current_direction != DMA_NONE;
+	}
+	dma_direction dma_dir;
+
+	// State-related functions
+	void step(bool timeout);
+	void check_irq();
+	void check_drq();
+	void reset_disconnect();
+	void send_byte(scsi_data_target data_source);
+	void recv_byte();
+	void function_complete();
+	void function_bus_complete();
+	void bus_complete();
+	void dma_set(dma_direction dir);
+	void decrement_tcounter(int count = 1);
+	bool transfer_count_zero();
+	void delay(int cycles);
+	void delay_cycles(int cycles);
+	void arbitrate();
+	void clear_fifo();
+	void auto_phase_transfer(int new_phase);
+
+	// AUXCTRL constants and functions
+	const uint32_t AUXCTRL_DMAEDGE = 0x04;
+	const uint32_t AUXCTRL_SETRST = 0x20;
+	const uint32_t AUXCTRL_CRST = 0x40;
+	const uint32_t AUXCTRL_SRST = 0x80;
+	uint32_t auxctrl_r();
+	void auxctrl_w(uint32_t data);
+
+	// FIFOCTRL constants and functions
+	// Based on the existence of CLREVEN/ODD, the fact that NetBSD only uses EVEN, and the max is 8
+	// even though this is a 4 bit value, it seems likely that there are actually two FIFOs,
+	// one in the even slots, and one in the odd slots
+	const uint32_t FIFOC_FSLOT = 0x0f; // Free slots in FIFO, max 8. Free slots = 8 - (FIFOCTRL & FIFOC_FSLOT)
+	const uint32_t FIFOC_SSTKACT = 0x10;
+	const uint32_t FIFOC_RQOVRN = 0x20;
+	const uint32_t FIFOC_CLREVEN = 0x00;
+	const uint32_t FIFOC_CLRODD = 0x40;
+	const uint32_t FIFOC_FLUSH = 0x80;
+	const uint32_t FIFOC_LOAD = 0xc0;
+	uint32_t fifoctrl_r();
+	void fifoctrl_w(uint32_t data);
+	std::queue<uint32_t> m_even_fifo;
+	std::queue<uint32_t> m_odd_fifo;
+	void clear_queue(std::queue<uint32_t>& queue)
+	{
+		while (!queue.empty())
+		{
+			queue.pop();
+		}
+	}
+
+	// spstat - not fully implemented yet
+	const uint32_t SPS_IDLE = 0x00;
+	const uint32_t SPS_MSGOUT = 0x04;
+	const uint32_t SPS_COMMAND = 0x05;
+	const uint32_t SPS_INTR = 0x08;
+	const uint32_t SPS_STATUS = 0x0c;
+	const uint32_t SPS_MSGIN = 0x0d;
+	const uint32_t SPS_DATAOUT = 0x0e;
+	const uint32_t SPS_DATAIN = 0x0f;
+	uint32_t spstat_r();
+	uint32_t prcmd_to_spstat(uint32_t cmd)
+	{
+		uint32_t spstat_val = 0;
+		switch (cmd)
+		{
+			case PRC_DATAIN:
+				spstat_val = SPS_DATAIN;
+				break;
+			case PRC_DATAOUT:
+				spstat_val = SPS_DATAOUT;
+				break;
+			case PRC_COMMAND:
+				spstat_val = SPS_COMMAND;
+				break;
+			case PRC_STATUS:
+				spstat_val = SPS_STATUS;
+				break;
+			case PRC_MSGOUT:
+				spstat_val = SPS_MSGOUT;
+				break;
+			case PRC_MSGIN:
+				spstat_val = SPS_MSGIN;
+				break;
+		}
+		return spstat_val;
+	}
```

Comment:
These kinds of things are probably better to do as inline functions in the .cpp file to make the header less fragile.